homeassistant/components/mqtt/config_flow.py

```diff
@@ -206,15 +226,164 @@ def async_get_options_flow(
 
+    async def _async_install_addon(self) -> None:
+        """Install the Mosquitto Mqtt broker add-on."""
+        addon_manager: AddonManager = get_addon_manager(self.hass)
+        await addon_manager.async_schedule_install_addon()
+
+    async def async_step_install_failed(
+        self, user_input: dict[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """Add-on installation failed."""
+        return self.async_abort(
+            reason="addon_install_failed",
+            description_placeholders={"addon": self._addon_manager.addon_name},
+        )
+
+    async def async_step_install_addon(
+        self, user_input: dict[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """Install Mosquitto Broker add-on."""
+        if self.install_task is None:
+            self.install_task = self.hass.async_create_task(self._async_install_addon())
+
+        if not self.install_task.done():
+            return self.async_show_progress(
+                step_id="install_addon",
+                progress_action="install_addon",
+                progress_task=self.install_task,
+            )
+
+        try:
+            await self.install_task
+        except AddonError as err:
+            _LOGGER.error(err)
+            return self.async_show_progress_done(next_step_id="install_failed")
+        finally:
+            self.install_task = None
+
+        return self.async_show_progress_done(next_step_id="start_addon")
+
+    async def async_step_start_failed(
+        self, user_input: dict[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """Add-on start failed."""
+        return self.async_abort(
+            reason="addon_start_failed",
+            description_placeholders={"addon": self._addon_manager.addon_name},
+        )
+
+    async def async_step_start_addon(
+        self, user_input: dict[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """Start Mosquitto Broker add-on."""
+        if not self.start_task:
+            self.start_task = self.hass.async_create_task(self._async_start_addon())
+        if not self.start_task.done():
+            return self.async_show_progress(
+                step_id="start_addon",
+                progress_action="start_addon",
+                progress_task=self.start_task,
+            )
+        try:
+            await self.start_task
+        except (AddonError, AbortFlow) as err:
+            _LOGGER.error(err)
+            return self.async_show_progress_done(next_step_id="start_failed")
+        finally:
+            self.start_task = None
+
+        return self.async_show_progress_done(next_step_id="setup_entry_from_discovery")
+
+    async def _async_start_addon(self) -> None:
+        """Start the Mosquitto Broker add-on."""
+        addon_manager: AddonManager = get_addon_manager(self.hass)
+        await addon_manager.async_schedule_start_addon()
+
+        # Sleep some seconds to let the add-on start properly before connecting.
```

Comment:
Take a look at  this class: <https://github.com/home-assistant/core/blob/828b6780864c707cb76aad2681772d664fd89263/homeassistant/components/homeassistant_hardware/silabs_multiprotocol_addon.py#L76>

It moved the waiting to the add-on manager which simplifies the caller's logic.

Can that class be used here too? `matter` and `zwave_js` should probably also use that class.

Reply:
I have looked at this class. And it can help a little to install the addon. But not for the progress code.
To start the add on the class does not check if the discovery info is available and if the broker can be connected with the discovery data. It will take some more time after starting the addon befoe the discovery info comes available. So I'd rather not the `WaitingAddonManager` class at this state, as it does not really help, and seems to make things more complicated.